File: src/core/ops/wiring.ts
import type { Registry } from "../../registry/types.ts";
import { MenvError } from "../errors.ts";
import type { ValueRecord } from "../refs.ts";
import { findDependents } from "../refs.ts";
import type { OpResult } from "./util.ts";
import { cloneRegistry, newPlan, requireConsumer, requireVariable, requireVault } from "./util.ts";

export interface WireInput {
  name: string;
  vault: string;
  consumers: string[];
  shared?: boolean;
  key?: string;
  newKey: () => string; // injected (crypto.randomUUID in production) for deterministic tests
  // With --key, an already-wired consumer is RE-KEYED onto that existing key
  // (rather than erroring), so it shares storage with the key's other holders.
  // A re-key can vacate the consumer's old key; if no consumer of this variable
  // still references it, it is orphaned — removed only when removeOrphans is set
  // and the vault is openable, otherwise kept with an ORPHANED_KEYS warning.
  removeOrphans?: boolean;
  openable?: Set<string>;
}

export function planWire(registry: Registry, input: WireInput): OpResult {
  requireVariable(registry, input.name);
  requireVault(registry, input.vault);
  for (const c of input.consumers) requireConsumer(registry, c);
  if (input.shared === true && input.key !== undefined) {
    throw new MenvError("VALIDATION", "--shared and --key are mutually exclusive");
  }
  const existing = registry.variables[input.name]?.vaultMapping[input.vault] ?? {};
  // Re-keying an already-wired consumer is only allowed in --key (join) mode;
  // the fresh/shared modes still demand an explicit unwire first.
  const rekeying = input.key !== undefined;
  const already = input.consumers.filter((c) => existing[c] !== undefined);
  if (!rekeying && already.length > 0) {
    throw new MenvError(
      "VALIDATION",
      `"${input.name}" is already wired to ${already.join(", ")} in vault "${input.vault}" (unwire first)`,
    );
  }
  const sharedKey = input.key ?? (input.shared === true ? input.newKey() : undefined);
  const next = cloneRegistry(registry);
  const def = next.variables[input.name];
  const plan = newPlan();
  if (def !== undefined) {
    const mapping = def.vaultMapping[input.vault] ?? {};
    const vacated = new Set<string>();
    for (const c of input.consumers) {
      const prev = mapping[c];
      const key = sharedKey ?? input.newKey();
      if (prev !== undefined) {
        if (prev.key === key) continue; // already on the target key — nothing to do
        vacated.add(prev.key);
        mapping[c] = { key, ...(prev.disabled === true ? { disabled: true } : {}) }; // preserve disabled
        plan.registry.push({
          action: "set",
          path: `variables.${input.name}.vaultMapping.${input.vault}.${c}.key`,
          summary: `re-key "${input.name}" → "${c}" to share key (vault "${input.vault}")`,
        });
      } else {
        mapping[c] = { key };
        plan.registry.push({
          action: "set",
          path: `variables.${input.name}.vaultMapping.${input.vault}.${c}`,
          summary: `wire "${input.name}" → "${c}" (vault "${input.vault}")`,
        });
      }
    }
    def.vaultMapping[input.vault] = mapping;
    const surviving = new Set(Object.values(mapping).map((e) => e.key));
    for (const key of [...vacated].sort()) {
      if (surviving.has(key)) continue;
      collectOrphan(plan, input.vault, key, input.name, input.removeOrphans === true, input.openable);
    }
  }
  return { next, plan };
}

// Shared orphaned-key policy: drop it only when removeOrphans is set and the
// vault can be opened; otherwise leave it and surface an ORPHANED_KEYS warning.
function collectOrphan(
  plan: OpResult["plan"],
  vault: string,
  key: string,
  name: string,
  removeOrphans: boolean,
  openable: Set<string> | undefined,
): void {
  if (!removeOrphans) {
    plan.warnings.push({
      code: "ORPHANED_KEYS",
      message: `key "${key}" for "${name}" is now unused in vault "${vault}" — left in place (use --remove-orphans to drop it)`,
    });
  } else if (openable?.has(vault) === true) {
    plan.vaults.push({ vault, action: "remove", key });
  } else {
    plan.warnings.push({
      code: "ORPHANED_KEYS",
      message: `vault "${vault}" could not be opened — orphaned key "${key}" remains (menv check will report it)`,
    });
  }
}

export interface UnwireInput {
  name: string;
  vault: string;
  consumers: string[];
  records: ValueRecord[]; // pre-collected from `vault`
  unverified: string[];
  openable: Set<string>;
  // Orphaned keys (no surviving consumer of this variable) are dropped only when
  // set and the vault is openable; otherwise kept with an ORPHANED_KEYS warning.
  removeOrphans?: boolean;
}

export function planUnwire(registry: Registry, input: UnwireInput): OpResult {
  const def = requireVariable(registry, input.name);
  requireVault(registry, input.vault);
  const mapping = def.vaultMapping[input.vault] ?? {};
  const missing = input.consumers.filter((c) => mapping[c] === undefined);
  if (missing.length > 0) {
    throw new MenvError(
      "VALIDATION",
      `"${input.name}" is not wired to ${missing.join(", ")} in vault "${input.vault}"`,
    );
  }

  const plan = newPlan();
  const dependents = findDependents(
    input.name,
    input.records.filter(
      (r) => r.vault === input.vault && input.consumers.includes(r.consumer) && r.variable !== input.name,
    ),
  );
  for (const d of dependents) {
    plan.blockers.push({
      code: "DEPENDENT_REFERENCE",
      message: `"${d.variable}" references \${${input.name}} (vault "${d.vault}", consumer "${d.consumer}")`,
    });
  }
  for (const v of input.unverified.filter((u) => u === input.vault)) {
    plan.blockers.push({
      code: "UNVERIFIED_REFERENCES",
      message: `vault "${v}" could not be opened — references to \${${input.name}} there are unverified`,
    });
  }

  const next = cloneRegistry(registry);
  const nextDef = next.variables[input.name];
  if (nextDef !== undefined) {
    const nextMapping = nextDef.vaultMapping[input.vault] ?? {};
    const removedKeys = new Set<string>();
    for (const c of input.consumers) {
      const entry = nextMapping[c];
      if (entry !== undefined) removedKeys.add(entry.key);
      delete nextMapping[c];
      plan.registry.push({
        action: "remove",
        path: `variables.${input.name}.vaultMapping.${input.vault}.${c}`,
        summary: `unwire "${input.name}" from "${c}" (vault "${input.vault}")`,
      });
    }
    const survivingKeys = new Set(Object.values(nextMapping).map((e) => e.key));
    for (const key of [...removedKeys].sort()) {
      if (survivingKeys.has(key)) continue;
      collectOrphan(plan, input.vault, key, input.name, input.removeOrphans === true, input.openable);
    }
    if (Object.keys(nextMapping).length === 0) delete nextDef.vaultMapping[input.vault];
  }
  return { next, plan };
}

export interface SetDisabledInput {
  name: string;
  vault: string;
  consumer: string;
  disabled: boolean;
}

export function planSetDisabled(registry: Registry, input: SetDisabledInput): OpResult {
  const def = requireVariable(registry, input.name);
  requireVault(registry, input.vault);
  requireConsumer(registry, input.consumer);
  const entry = def.vaultMapping[input.vault]?.[input.consumer];
  if (entry === undefined) {
    throw new MenvError(
      "NOT_FOUND",
      `"${input.name}" is not wired to "${input.consumer}" in vault "${input.vault}"`,
    );
  }
  const plan = newPlan();
  const current = entry.disabled === true;
  if (current === input.disabled) {
    plan.warnings.push({
      code: "NOOP",
      message: `"${input.name}" is already ${input.disabled ? "disabled" : "enabled"} for "${input.consumer}" in vault "${input.vault}"`,
    });
    return { next: registry, plan };
  }
  const next = cloneRegistry(registry);
  const nextEntry = next.variables[input.name]?.vaultMapping[input.vault]?.[input.consumer];
  if (nextEntry !== undefined) {
    if (input.disabled) nextEntry.disabled = true;
    else delete nextEntry.disabled;
  }
  plan.registry.push({
    action: "set",
    path: `variables.${input.name}.vaultMapping.${input.vault}.${input.consumer}.disabled`,
    summary: `${input.disabled ? "disable" : "enable"} "${input.name}" for "${input.consumer}" (vault "${input.vault}")`,
  });
  return { next, plan };
}

File: go/internal/core/ops/wiring.go
package ops

import (
	"fmt"
	"sort"

	"github.com/nikrabaev/menv/internal/core"
	"github.com/nikrabaev/menv/internal/registry"
)

// WireInput holds the parameters for PlanWire.
type WireInput struct {
	Name          string
	Vault         string
	Consumers     []string
	Shared        bool
	Key           string   // non-empty = re-key consumers onto this existing key
	NewKey        func() string
	RemoveOrphans bool
	Openable      map[string]bool
}

// PlanWire maps a variable to a vault key for the given consumers.
func PlanWire(r registry.Registry, input WireInput) (OpResult, error) {
	if _, err := RequireVariable(r, input.Name); err != nil {
		return OpResult{}, err
	}
	if _, err := RequireVault(r, input.Vault); err != nil {
		return OpResult{}, err
	}
	for _, c := range input.Consumers {
		if _, err := RequireConsumer(r, c); err != nil {
			return OpResult{}, err
		}
	}
	if input.Shared && input.Key != "" {
		return OpResult{}, &core.MenvError{
			Code:    core.ErrValidation,
			Message: "--shared and --key are mutually exclusive",
		}
	}

	existing := map[string]registry.MappingEntry{}
	if m, ok := r.Variables[input.Name].VaultMapping[input.Vault]; ok {
		for k, v := range m {
			existing[k] = v
		}
	}

	rekeying := input.Key != ""
	var already []string
	for _, c := range input.Consumers {
		if _, ok := existing[c]; ok {
			already = append(already, c)
		}
	}
	if !rekeying && len(already) > 0 {
		sort.Strings(already)
		return OpResult{}, &core.MenvError{
			Code:    core.ErrValidation,
			Message: fmt.Sprintf("%q is already wired to %s in vault %q (unwire first)", input.Name, joinComma(already), input.Vault),
		}
	}

	// Determine the shared key (if any).
	var sharedKey string
	if input.Key != "" {
		sharedKey = input.Key
	} else if input.Shared {
		sharedKey = input.NewKey()
	}

	next := CloneRegistry(r)
	def := next.Variables[input.Name]
	if def.VaultMapping == nil {
		def.VaultMapping = map[string]map[string]registry.MappingEntry{}
	}
	if def.VaultMapping[input.Vault] == nil {
		def.VaultMapping[input.Vault] = map[string]registry.MappingEntry{}
	}
	mapping := def.VaultMapping[input.Vault]

	plan := NewPlan()
	vacated := map[string]bool{}

	for _, c := range input.Consumers {
		key := sharedKey
		if key == "" {
			key = input.NewKey()
		}
		prev, hasPrev := mapping[c]
		if hasPrev {
			if prev.Key == key {
				continue // already on target key
			}
			vacated[prev.Key] = true
			entry := registry.MappingEntry{Key: key}
			if prev.Disabled {
				entry.Disabled = true
			}
			mapping[c] = entry
			plan.Registry = append(plan.Registry, core.RegistryOp{
				Action:  "set",
				Path:    fmt.Sprintf("variables.%s.vaultMapping.%s.%s.key", input.Name, input.Vault, c),
				Summary: fmt.Sprintf("re-key %q → %q to share key (vault %q)", input.Name, c, input.Vault),
			})
		} else {
			mapping[c] = registry.MappingEntry{Key: key}
			plan.Registry = append(plan.Registry, core.RegistryOp{
				Action:  "set",
				Path:    fmt.Sprintf("variables.%s.vaultMapping.%s.%s", input.Name, input.Vault, c),
				Summary: fmt.Sprintf("wire %q → %q (vault %q)", input.Name, c, input.Vault),
			})
		}
	}
	def.VaultMapping[input.Vault] = mapping
	next.Variables[input.Name] = def

	// Orphan handling for vacated keys.
	surviving := map[string]bool{}
	for _, e := range mapping {
		surviving[e.Key] = true
	}
	vacatedKeys := make([]string, 0, len(vacated))
	for k := range vacated {
		vacatedKeys = append(vacatedKeys, k)
	}
	sort.Strings(vacatedKeys)
	for _, key := range vacatedKeys {
		if surviving[key] {
			continue
		}
		collectOrphan(&plan, input.Vault, key, input.Name, input.RemoveOrphans, input.Openable)
	}

	return OpResult{Next: next, Plan: plan}, nil
}

// UnwireInput holds the parameters for PlanUnwire.
type UnwireInput struct {
	Name          string
	Vault         string
	Consumers     []string
	Records       []core.ValueRecord
	Unverified    []string
	Openable      map[string]bool
	RemoveOrphans bool
}

// PlanUnwire removes consumer mappings from a variable in a vault.
func PlanUnwire(r registry.Registry, input UnwireInput) (OpResult, error) {
	def, err := RequireVariable(r, input.Name)
	if err != nil {
		return OpResult{}, err
	}
	if _, err := RequireVault(r, input.Vault); err != nil {
		return OpResult{}, err
	}
	mapping := def.VaultMapping[input.Vault]
	var missing []string
	for _, c := range input.Consumers {
		if _, ok := mapping[c]; !ok {
			missing = append(missing, c)
		}
	}
	if len(missing) > 0 {
		sort.Strings(missing)
		return OpResult{}, &core.MenvError{
			Code:    core.ErrValidation,
			Message: fmt.Sprintf("%q is not wired to %s in vault %q", input.Name, joinComma(missing), input.Vault),
		}
	}

	plan := NewPlan()
	consumerSet := map[string]bool{}
	for _, c := range input.Consumers {
		consumerSet[c] = true
	}

	// Dependency check within the affected (vault, consumers).
	affectedRecords := make([]core.ValueRecord, 0, len(input.Records))
	for _, rec := range input.Records {
		if rec.Vault == input.Vault && consumerSet[rec.Consumer] && rec.Variable != input.Name {
			affectedRecords = append(affectedRecords, rec)
		}
	}
	dependents := core.FindDependents(input.Name, affectedRecords)
	for _, d := range dependents {
		plan.Blockers = append(plan.Blockers, core.PlanIssue{
			Code:    "DEPENDENT_REFERENCE",
			Message: fmt.Sprintf("%q references ${%s} (vault %q, consumer %q)", d.Variable, input.Name, d.Vault, d.Consumer),
		})
	}
	for _, v := range input.Unverified {
		if v == input.Vault {
			plan.Blockers = append(plan.Blockers, core.PlanIssue{
				Code:    "UNVERIFIED_REFERENCES",
				Message: fmt.Sprintf("vault %q could not be opened — references to ${%s} there are unverified", v, input.Name),
			})
		}
	}

	next := CloneRegistry(r)
	nextDef := next.Variables[input.Name]
	if nextDef.VaultMapping == nil {
		nextDef.VaultMapping = map[string]map[string]registry.MappingEntry{}
	}
	nextMapping := nextDef.VaultMapping[input.Vault]
	if nextMapping == nil {
		nextMapping = map[string]registry.MappingEntry{}
	}

	removedKeys := map[string]bool{}
	for _, c := range input.Consumers {
		if entry, ok := nextMapping[c]; ok {
			removedKeys[entry.Key] = true
		}
		delete(nextMapping, c)
		plan.Registry = append(plan.Registry, core.RegistryOp{
			Action:  "remove",
			Path:    fmt.Sprintf("variables.%s.vaultMapping.%s.%s", input.Name, input.Vault, c),
			Summary: fmt.Sprintf("unwire %q from %q (vault %q)", input.Name, c, input.Vault),
		})
	}

	surviving := map[string]bool{}
	for _, e := range nextMapping {
		surviving[e.Key] = true
	}
	removedKeyList := make([]string, 0, len(removedKeys))
	for k := range removedKeys {
		removedKeyList = append(removedKeyList, k)
	}
	sort.Strings(removedKeyList)
	for _, key := range removedKeyList {
		if surviving[key] {
			continue
		}
		collectOrphan(&plan, input.Vault, key, input.Name, input.RemoveOrphans, input.Openable)
	}

	if len(nextMapping) == 0 {
		delete(nextDef.VaultMapping, input.Vault)
	} else {
		nextDef.VaultMapping[input.Vault] = nextMapping
	}
	next.Variables[input.Name] = nextDef

	return OpResult{Next: next, Plan: plan}, nil
}

// SetDisabledInput holds the parameters for PlanSetDisabled.
type SetDisabledInput struct {
	Name     string
	Vault    string
	Consumer string
	Disabled bool
}

// PlanSetDisabled enables or disables a wired variable for a given consumer.
func PlanSetDisabled(r registry.Registry, input SetDisabledInput) (OpResult, error) {
	def, err := RequireVariable(r, input.Name)
	if err != nil {
		return OpResult{}, err
	}
	if _, err := RequireVault(r, input.Vault); err != nil {
		return OpResult{}, err
	}
	if _, err := RequireConsumer(r, input.Consumer); err != nil {
		return OpResult{}, err
	}
	entry, ok := def.VaultMapping[input.Vault][input.Consumer]
	if !ok {
		return OpResult{}, &core.MenvError{
			Code:    core.ErrNotFound,
			Message: fmt.Sprintf("%q is not wired to %q in vault %q", input.Name, input.Consumer, input.Vault),
		}
	}
	plan := NewPlan()
	if entry.Disabled == input.Disabled {
		plan.Warnings = append(plan.Warnings, core.PlanIssue{
			Code:    "NOOP",
			Message: fmt.Sprintf("%q is already %s for %q in vault %q", input.Name, disabledStr(input.Disabled), input.Consumer, input.Vault),
		})
		return OpResult{Next: r, Plan: plan}, nil
	}
	next := CloneRegistry(r)
	nextEntry := next.Variables[input.Name].VaultMapping[input.Vault][input.Consumer]
	nextEntry.Disabled = input.Disabled
	next.Variables[input.Name].VaultMapping[input.Vault][input.Consumer] = nextEntry
	plan.Registry = append(plan.Registry, core.RegistryOp{
		Action:  "set",
		Path:    fmt.Sprintf("variables.%s.vaultMapping.%s.%s.disabled", input.Name, input.Vault, input.Consumer),
		Summary: fmt.Sprintf("%s %q for %q (vault %q)", disabledStr(input.Disabled), input.Name, input.Consumer, input.Vault),
	})
	return OpResult{Next: next, Plan: plan}, nil
}

func disabledStr(d bool) string {
	if d {
		return "disable"
	}
	return "enable"
}

func collectOrphan(plan *core.Plan, vault, key, name string, removeOrphans bool, openable map[string]bool) {
	if !removeOrphans {
		plan.Warnings = append(plan.Warnings, core.PlanIssue{
			Code:    "ORPHANED_KEYS",
			Message: fmt.Sprintf("key %q for %q is now unused in vault %q — left in place (use --remove-orphans to drop it)", key, name, vault),
		})
		return
	}
	if openable[vault] {
		plan.Vaults = append(plan.Vaults, core.VaultOp{Vault: vault, Action: "remove", Key: key})
	} else {
		plan.Warnings = append(plan.Warnings, core.PlanIssue{
			Code:    "ORPHANED_KEYS",
			Message: fmt.Sprintf("vault %q could not be opened — orphaned key %q remains (menv check will report it)", vault, key),
		})
	}
}
